src/messages/tests/integration.rs

```diff
@@ -0,0 +1,749 @@
+    assert_eq!(validate_message, Err(MsgErr::ReceiverMustBeSigner));
+
+    let participants = vec![setup.signer1];
+    let mut share_commitment = generate_share_commitment(&shares, participants);
+
+    // change the payload to have only 1 commitment
+    let payload = Payload::SharePackage(SharePackage {
+        group_public,
+        secret_share: secret_share,
+        share_commitment: share_commitment.clone(),
+    });
+    let validate_payload = Validate::validate(&payload);
+    assert_eq!(
+        validate_payload,
+        Err(MsgErr::NotEnoughCommitments(constants::MIN_SIGNERS))
+    );
+
+    // build and use too many commitments
+    for i in 2..constants::MAX_SIGNERS as u64 + 2 {
+        share_commitment.insert(
+            ParticipantId::Signer(i),
+            share_commitment.clone()[&setup.signer1],
+        );
+    }
+    let payload = Payload::SharePackage(SharePackage {
+        group_public,
+        secret_share,
+        share_commitment,
+    });
+    let validate_payload = Validate::validate(&payload);
+    assert_eq!(validate_payload, Err(MsgErr::TooManyCommitments));
+}
+
+#[test]
+fn serialize_sharepackage() {
+    let setup = basic_setup();
+
+    let (shares, _pubkeys) =
+        frost::keygen_with_dealer(setup.num_signers, setup.threshold, setup.rng.clone()).unwrap();
+
+    let header = create_valid_header(setup.dealer, setup.signer1);
+
+    let group_public = VerificationKey::from(
+        verification_key::VerificationKey::try_from(shares[0].group_public.bytes).unwrap(),
+    );
+    let secret_share = Secret(shares[0].share.value.0.to_bytes());
+
+    let participants = vec![setup.signer1];
+    let share_commitment = generate_share_commitment(&shares, participants);
+
+    let payload = Payload::SharePackage(SharePackage {
+        group_public,
+        secret_share,
+        share_commitment: share_commitment.clone(),
+    });
+
+    let message = Message {
+        header: header,
+        payload: payload.clone(),
+    };
+
+    // check general structure and header serialization/deserialization
+    serialize_message(message, setup.dealer, setup.signer1);
+
+    // check payload serialization/deserialization
+    let mut payload_serialized_bytes = bincode::serialize(&payload).unwrap();
+    // remove the padding. TODO: why is this happening ?
+    payload_serialized_bytes =
+        (&payload_serialized_bytes[4..payload_serialized_bytes.len()]).to_vec();
+
+    // group_public is 32 bytes
+    let deserialized_group_public: VerificationKey =
+        bincode::deserialize(&payload_serialized_bytes[0..32]).unwrap();
+    // secret share is 32 bytes
+    let deserialized_secret_share: Secret =
+        bincode::deserialize(&payload_serialized_bytes[32..64]).unwrap();
+    // rest of the message is the map
+    let deserialized_share_commitment: BTreeMap<ParticipantId, Commitment> =
+        bincode::deserialize(&payload_serialized_bytes[64..payload_serialized_bytes.len()])
+            .unwrap();
```

Comment:
One thing of note is that I think `bincode` will serialize the `BTreeMap` with an 8-byte length (`u64`), which doesn't match exactly the specification in the RFC.

Not sure if there's any usefulness in adding an `assert` for the deserialized length and for the remaining length. Like: 
```
let expected_map_length = (8 + 32 + 32) * 1;
assert_eq!(payload_serialized_bytes[64], expected_map_length);
assert_eq!(payload_serialized_bytes.len() - 64, 1 + expected_map_length);
```

Reply:
I opened a a new ticket to take action in the bincode serialization/deserialization issues of the `BTreeMap`: https://github.com/ZcashFoundation/redjubjub/issues/118